src/Symfony/Component/Routing/Loader/ServiceRouterLoader.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Routing\Loader;
+
+use Symfony\Component\Config\Loader\Loader;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Routing\RouteCollection;
+
+/**
+ * A route loader that executes a service to load the routes.
+ *
+ * The depends on the DependencyInjection component
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class ServiceRouterLoader extends Loader
+{
+    /**
+     * @var ContainerInterface
+     */
+    private $container;
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * Calls the service that will load the routes.
+     *
+     * @param string      $resource The name of the service to load
+     * @param string|null $type     The resource type
+     *
+     * @return RouteCollection
+     */
+    public function load($resource, $type = null)
+    {
+        $service = $this->container->get($resource);
+
+        if (!$service instanceof RouteLoaderInterface) {
+            throw new \LogicException(sprintf('Service "%s" must implement RouteLoaderInterface.', $resource));
+        }
+
+        $routeCollection = $service->getRouteCollection($this);
+
+        // make the service file tracked so that if it changes, the cache rebuilds
+        $this->addClassResource(new \ReflectionClass($service), $routeCollection);
```

Comment:
it is indeed the job of the loader to register resources, as the loader knows how it loaded routes.

For routes loaded from the database, you would have to implement a custom resource for that. and it would be a pain until #7230 is implemented.
However, a database loader is probably a very bad idea: the resource-based invalidation of the cache is done only in debug-mode, and you should never put the production website in debug mode (unless you don't care about performance at all of course, but you could as well disable the routing cache to simplify your life if you don't care). So you would have to clear your prod cache explicitly each time the DB changes. This is exactly why the CMF project does not implement a database loader but implements a separate dynamic router instead. This way, the cached router contains only static routes, not dynamic ones, and there is no need to invalidate the routing cache on each DB change.